[ao][ns] PNP demo for exposing arbitrary model transforms

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89892

Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

HDCharles committed:
[ao][ns] PNP demo for exposing arbitrary model transforms

Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

vkuzo commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -752,6 +752,9 @@ def prepare_n_shadows_model(
     example_inputs: Any,
     qconfig_multi_mapping: QConfigMultiMapping,
     backend_config: BackendConfig,
+    exposed_prepare_function: Any=None,
```
nit: fix lint

vkuzo replied:
would a name such as `prepare_fn_override` and `convert_fn_override`, or `custom_prepare_fn` / `custom_convert_fn` make more sense?  the `exposed` word seems ambiguous

HDCharles replied:
ok, will update

vkuzo commented on torch/ao/ns/_numeric_suite_fx.py:
```diff
@@ -862,7 +870,7 @@ def loggers_set_save_activations(
             child.save_activations = save_activations
 
-def convert_n_shadows_model(model: GraphModule) -> GraphModule:
+def convert_n_shadows_model(model: GraphModule, exposed_convert_function: Any=None, exposed_convert_kwargs: Any=None) -> GraphModule:
```
can we use `Callable` and `Dict[str, Any]` as the respective types?

HDCharles replied:
fixed

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

vkuzo commented on torch/ao/ns/fx/n_shadows_utils.py:
```diff
@@ -566,9 +568,19 @@ def handle_subgraph_candidate(
+        else:
+            if custom_prepare_kwargs is None:
+                custom_prepare_kwargs = {}
+            custom_prepare_kwargs["example_inputs"] = example_inputs
```
should we check that the original `custom_prepare_kwargs` don't specify `example_inputs`, `prepare_custom_config`, `qconfig_mapping`?

HDCharles replied:
yeah, that makes sense

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao][ns] PNP demo for exposing arbitrary model transforms"


Adding way to use arbitrary prepare and convert functions with PNP.

Differential Revision: [D41167110](https://our.internmc.facebook.com/intern/diff/D41167110/)

[ghstack-poisoned]

mehtanirav commented:
@pytorchbot rebase
